DatadogInternal/Sources/NetworkInstrumentation/URLSession/URLSessionDataDelegateSwizzler.swift

```diff
@@ -0,0 +1,100 @@
+
+/// Swizzles `URLSessionDataDelegate` callbacks.
+internal class URLSessionDataDelegateSwizzler {
+    static var didReceiveMap: [String: DidReceive?] = [:]
+    static var lock = NSRecursiveLock()
+
+    static var isBinded: Bool {
+        lock.lock()
+        defer { lock.unlock() }
+        return didReceiveMap.isEmpty == false
+    }
+
+    static func bindIfNeeded(delegateClass: AnyClass, interceptDidReceive: @escaping (URLSession, URLSessionDataTask, Data) -> Void) throws {
+        lock.lock()
+        defer { lock.unlock() }
+
+        guard isBinded == false else {
+            return
+        }
+
+        try bind(delegateClass: delegateClass, interceptDidReceive: interceptDidReceive)
+    }
+
+    static func bind(delegateClass: AnyClass, interceptDidReceive: @escaping (URLSession, URLSessionDataTask, Data) -> Void) throws {
+        lock.lock()
+        defer { lock.unlock() }
+
+        let didReceive = try DidReceive.build(klass: delegateClass)
+
+        didReceive.swizzle(intercept: interceptDidReceive)
+        didReceiveMap[key(from: delegateClass)] = didReceive
+    }
+
+    static func unbind(delegateClass: AnyClass) {
+        lock.lock()
+        defer { lock.unlock() }
+
+        guard let didReceive = didReceiveMap[key(from: delegateClass)] else {
+            return
+        }
+
+        didReceive?.unswizzle()
+        didReceiveMap[key(from: delegateClass)] = nil
+    }
+
+    static func unbindAll() {
+        lock.lock()
+        defer { lock.unlock() }
+
+        didReceiveMap.forEach { _, didReceive in
+            didReceive?.unswizzle()
+        }
+        didReceiveMap.removeAll()
+    }
+
+    /// Swizzles `urlSession(_:dataTask:didReceive:)` callback.
+    /// This callback is called when the response is received.
+    /// It is called multiple times for a single request, each time with a new chunk of data.
+    class DidReceive: MethodSwizzler<@convention(c) (URLSessionDataDelegate, Selector, URLSession, URLSessionDataTask, Data) -> Void, @convention(block) (URLSessionDataDelegate, URLSession, URLSessionDataTask, Data) -> Void> {
+        private static let selector = #selector(URLSessionDataDelegate.urlSession(_:dataTask:didReceive:))
+
+        private let method: FoundMethod
+
+        static func build(klass: AnyClass) throws -> DidReceive {
+            return try DidReceive(selector: self.selector, klass: klass)
+        }
+
+        private init(selector: Selector, klass: AnyClass) throws {
+            do {
+                method = try Self.findMethod(with: selector, in: klass)
+            } catch {
+                // URLSessionDataDelegate doesn't implement the selector, so we inject it and swizzle it
+                let block: @convention(block) (URLSessionDataDelegate, URLSession, URLSessionDataTask, Data) -> Void = { delegate, session, task, data in
+                }
+                let imp = imp_implementationWithBlock(block)
+                class_addMethod(klass, selector, imp, "@@@")
```

Comment:
question/ What is `"@@@"`? 🤔💭 If it's the method name, I feel it's better to include "datadog" in the name - that might help to debug and understand their stack traces.

Reply:
just updated it, i hope it will make it easier to understand